File: backend/library.py
"""
卫星数据文件管理 REST 路由：
    GET  /api/library/meta       列出可用数据源组 + 本地文件是否已下载/条数/时间
    POST /api/library/download   下载某数据源组原始文件到本地目录（同步返回）
    GET  /api/library/entries    浏览已下载原始文件解析出的卫星数据（支持名称/NORAD 搜索、按来源过滤）
"""

from __future__ import annotations

from datetime import datetime, timezone

from fastapi import APIRouter

from exceptions import NotFoundError, ValidationError  # noqa: E402
from lib import (  # noqa: E402
    download_group,
    find_entry_by_norad,
    get_satellite_info,
    list_all_entries,
    list_categories,
    list_downloaded,
    list_groups,
    read_group_entries,
)
import store  # noqa: E402
from tle import _parse_tle_fields  # noqa: E402

router = APIRouter()


def _group_status(g: dict, downloaded: dict) -> dict:
    """组定义 + 本地文件状态。"""
    local = downloaded.get(g["key"])
    return {
        "key": g["key"],
        "label": g["label"],
        "url": g["url"],
        "downloaded": bool(local),
        "count": local.get("count", 0) if local else 0,
        "size": local.get("size", 0) if local else 0,
        "fetched_at": local.get("fetched_at") if local else None,
    }


@router.get("/api/library/meta")
def library_meta():
    """返回按六大分类组织的数据源组 + 本地是否已下载文件（条数/大小/更新时间）。

    兼容：同时提供 categories（分类树）与 groups（平铺），供前端两种渲染方式选用。
    """
    downloaded = {d["key"]: d for d in list_downloaded()}
    categories = []
    groups = []
    for cat in list_categories():
        cat_groups = [_group_status(g, downloaded) for g in cat["groups"]]
        categories.append({"key": cat["key"], "label": cat["label"], "groups": cat_groups})
        groups.extend(cat_groups)
    total = sum(len(read_group_entries(d["key"]) or []) for d in downloaded.values())
    return {"categories": categories, "groups": groups, "total_entries": total}


@router.post("/api/library/download")
def library_download(payload: dict):
    """下载指定 CelesTrak 组原始文件到本地目录；同步返回；同组覆盖更新。"""
    key = str(payload.get("key", "")).strip()
    result = download_group(key)
    if result is None:
        valid_keys = [g["key"] for g in list_groups()]
        if key not in valid_keys:
            raise ValidationError(f"未知的数据源组: {key}")
        raise NotFoundError(f"数据源组下载失败: {key}（网络不可达或返回为空）")
    return result


@router.get("/api/library/entries")
def library_entries(q: str = "", source: str = ""):
    """浏览已下载原始文件解析出的卫星数据。

    q: 名称 / NORAD 号模糊匹配；source: 指定只读某组文件（留空 = 全部已下载组）。
    """
    if source:
        parsed = read_group_entries(source) or []
    else:
        parsed = list_all_entries()

    q = q.strip().lower()
    if q:
        if q.isdigit():
            parsed = [e for e in parsed if str(e["norad_id"]) == q or q in str(e["norad_id"])]
        else:
            parsed = [e for e in parsed if q in (e.get("name") or "").lower()]
    parsed.sort(key=lambda e: (e.get("name") or "").lower())
    return {
        "count": len(parsed),
        "generated_at": datetime.now(timezone.utc).isoformat(),
        "entries": parsed,
    }


@router.get("/api/library/detail")
def library_detail(norad_id: int):
    """返回库内指定卫星的详情：基础字段 + 从 TLE 解析的轨道根数（离线，默认全部可算）。"""
    entry = find_entry_by_norad(norad_id)
    if entry is None:
        raise NotFoundError(f"卫星库中未找到 NORAD {norad_id}（请先下载对应的数据源组）")
    orbit = _parse_tle_fields(entry.get("tle1", ""), entry.get("tle2", ""))
    return {
        "norad_id": entry["norad_id"],
        "name": entry.get("name", ""),
        "source": entry.get("source", ""),
        "tle1": entry.get("tle1", ""),
        "tle2": entry.get("tle2", ""),
        "tle_fetched_at": entry.get("tle_fetched_at", ""),
        "orbit": orbit,
    }


@router.get("/api/library/info")
def library_info(norad_id: int, refresh: bool = False):
    """返回库内指定卫星的档案信息（SatNOGS 基本信息 + AMSAT 频率）。

    优先返回本地缓存；refresh=True 强制联网刷新并更新缓存。
    未收档 / 查询失败返回 {found: False}，前端据此显示"该源暂无"。
    """
    info = get_satellite_info(norad_id, refresh=refresh)
    if info is None:
        return {
            "norad_id": norad_id,
            "found": False,
        }
    return {"norad_id": norad_id, "found": True, **info}


@router.post("/api/library/activate")
def library_activate(payload: dict):
    """把卫星库中的一颗卫星加入"已加入"列表（即配置页卫星管理表格）。

    该星必须已在本地卫星库（已下载对应数据源组）；已加入则报错。
    """
    norad = str(payload.get("norad_id", "")).strip()
    if not norad.isdigit():
        raise ValidationError("请输入有效的 NORAD 目录号（纯数字）")
    nid = int(norad)
    sats = store._load_satellites()
    if any(int(s["norad_id"]) == nid for s in sats):
        raise ValidationError(f"卫星（NORAD {nid}）已在已加入列表中")
    entry = find_entry_by_norad(nid)
    if entry is None:
        raise NotFoundError(f"卫星库中未找到 NORAD {nid}（请先下载对应的数据源组）")
    sat = {"id": str(nid), "name": entry.get("name", str(nid)), "norad_id": nid, "builtin": False}
    saved = store._save_satellites(sats + [sat])
    return {"satellites": saved, "satellite": sat}


@router.post("/api/library/deactivate")
def library_deactivate(payload: dict):
    """把某颗卫星从"已加入"列表移除（内置卫星不可删除）。"""
    sid = str(payload.get("id", "")).strip()
    if not sid:
        raise ValidationError("缺少卫星 id")
    sats = [s for s in store._load_satellites() if not (s["id"] == sid and not s.get("builtin"))]
    saved = store._save_satellites(sats)
    return {"satellites": saved}

File: backend-rs/src/routes/library.rs
//! 卫星数据文件管理 REST 路由：对应 backend/library.py。

use std::collections::HashMap;

use axum::Json;
use axum::extract::Query;
use chrono::Utc;
use serde::Deserialize;
use serde_json::{Value, json};

use crate::astroconv::iso_utc;
use crate::error::{ApiError, ApiResult};
use crate::libfiles;
use crate::store;
use crate::tle;

#[derive(Debug, Deserialize)]
pub struct EntriesQuery {
    #[serde(default)]
    pub q: String,
    #[serde(default)]
    pub source: String,
}

#[derive(Debug, Deserialize)]
pub struct DetailQuery {
    pub norad_id: i64,
}

#[derive(Debug, Deserialize)]
pub struct InfoQuery {
    pub norad_id: i64,
    #[serde(default)]
    pub refresh: bool,
}

/// 组定义 + 本地文件状态（对应 library._group_status）。
fn group_status(g: &Value, downloaded: &HashMap<String, Value>) -> Value {
    let key = g.get("key").and_then(|v| v.as_str()).unwrap_or("");
    match downloaded.get(key) {
        Some(local) => json!({
            "key": key,
            "label": g.get("label").cloned().unwrap_or(Value::Null),
            "url": g.get("url").cloned().unwrap_or(Value::Null),
            "downloaded": true,
            "count": local.get("count").cloned().unwrap_or(json!(0)),
            "size": local.get("size").cloned().unwrap_or(json!(0)),
            "fetched_at": local.get("fetched_at").cloned().unwrap_or(Value::Null),
        }),
        None => json!({
            "key": key,
            "label": g.get("label").cloned().unwrap_or(Value::Null),
            "url": g.get("url").cloned().unwrap_or(Value::Null),
            "downloaded": false,
            "count": 0,
            "size": 0,
            "fetched_at": Value::Null,
        }),
    }
}

/// GET /api/library/meta：数据源分类树 + 本地下载状态。
pub async fn meta() -> Json<Value> {
    let downloaded_list = libfiles::list_downloaded();
    let downloaded_map: HashMap<String, Value> = downloaded_list
        .into_iter()
        .filter_map(|d| {
            d.get("key")
                .and_then(|v| v.as_str())
                .map(|k| (k.to_string(), d.clone()))
        })
        .collect();

    let cats_in = libfiles::list_categories();
    let mut categories: Vec<Value> = Vec::new();
    let mut groups_flat: Vec<Value> = Vec::new();
    if let Some(cats) = cats_in.as_array() {
        for cat in cats {
            let mut cat_groups: Vec<Value> = Vec::new();
            if let Some(gs) = cat.get("groups").and_then(|v| v.as_array()) {
                for g in gs {
                    let status = group_status(g, &downloaded_map);
                    groups_flat.push(status.clone());
                    cat_groups.push(status);
                }
            }
            categories.push(json!({
                "key": cat.get("key").cloned().unwrap_or(Value::Null),
                "label": cat.get("label").cloned().unwrap_or(Value::Null),
                "groups": cat_groups,
            }));
        }
    }
    let total: usize = downloaded_map
        .keys()
        .map(|k| libfiles::read_group_entries(k).map(|e| e.len()).unwrap_or(0))
        .sum();
    Json(json!({
        "categories": categories,
        "groups": groups_flat,
        "total_entries": total,
    }))
}

/// POST /api/library/download：下载某 CelesTrak 组原始文件到本地目录。
pub async fn download(Json(payload): Json<Value>) -> ApiResult<Json<Value>> {
    let key = payload
        .get("key")
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .trim()
        .to_string();
    match libfiles::download_group(&key).await {
        Some(result) => Ok(Json(result)),
        None => {
            let valid = libfiles::list_groups().iter().any(|g| g.key == key);
            if !valid {
                Err(ApiError::validation(format!("未知的数据源组: {}", key)))
            } else {
                Err(ApiError::not_found(format!(
                    "数据源组下载失败: {}（网络不可达或返回为空）",
                    key
                )))
            }
        }
    }
}

/// GET /api/library/entries：浏览已下载数据源中的卫星（搜索/按来源过滤）。
pub async fn entries(Query(q): Query<EntriesQuery>) -> Json<Value> {
    let mut parsed = if q.source.trim().is_empty() {
        libfiles::list_all_entries()
    } else {
        libfiles::read_group_entries(q.source.trim()).unwrap_or_default()
    };
    let needle = q.q.trim().to_lowercase();
    if !needle.is_empty() {
        let digits = needle.chars().all(|c| c.is_ascii_digit());
        parsed.retain(|e| {
            if digits {
                let nid = e.get("norad_id").and_then(|v| v.as_i64()).map(|n| n.to_string());
                match nid {
                    Some(s) => s == needle || s.contains(&needle),
                    None => false,
                }
            } else {
                e.get("name")
                    .and_then(|v| v.as_str())
                    .map(|n| n.to_lowercase().contains(&needle))
                    .unwrap_or(false)
            }
        });
    }
    parsed.sort_by_key(|e| {
        e.get("name")
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .to_lowercase()
    });
    Json(json!({
        "count": parsed.len(),
        "generated_at": iso_utc(&Utc::now()),
        "entries": parsed,
    }))
}

/// GET /api/library/detail：库内指定卫星详情（TLE + 轨道根数）。
pub async fn detail(Query(q): Query<DetailQuery>) -> ApiResult<Json<Value>> {
    let entry = libfiles::find_entry_by_norad(q.norad_id).ok_or_else(|| {
        ApiError::not_found(format!(
            "卫星库中未找到 NORAD {}（请先下载对应的数据源组）",
            q.norad_id
        ))
    })?;
    let tle1 = entry.get("tle1").and_then(|v| v.as_str()).unwrap_or("");
    let tle2 = entry.get("tle2").and_then(|v| v.as_str()).unwrap_or("");
    let orbit = tle::parse_tle_fields(tle1, tle2);
    Ok(Json(json!({
        "norad_id": entry.get("norad_id").cloned().unwrap_or(Value::Null),
        "name": entry.get("name").cloned().unwrap_or(json!("")),
        "source": entry.get("source").cloned().unwrap_or(json!("")),
        "tle1": tle1,
        "tle2": tle2,
        "tle_fetched_at": entry.get("tle_fetched_at").cloned().unwrap_or(json!("")),
        "orbit": orbit,
    })))
}

/// GET /api/library/info：库内指定卫星档案（SatNOGS + AMSAT）。
pub async fn info(Query(q): Query<InfoQuery>) -> Json<Value> {
    match libfiles::get_satellite_info(q.norad_id, q.refresh).await {
        None => Json(json!({ "norad_id": q.norad_id, "found": false })),
        Some(info) => {
            let mut out = json!({ "norad_id": q.norad_id, "found": true });
            if let (Some(o), Some(i)) = (out.as_object_mut(), info.as_object()) {
                for (k, v) in i {
                    o.insert(k.clone(), v.clone());
                }
            }
            Json(out)
        }
    }
}

/// POST /api/library/activate：把库内卫星加入"已加入"列表。
pub async fn activate(Json(payload): Json<Value>) -> ApiResult<Json<Value>> {
    let norad = match payload.get("norad_id") {
        Some(Value::String(s)) => s.trim().to_string(),
        Some(Value::Number(n)) => n.as_i64().map(|i| i.to_string()).unwrap_or_default(),
        _ => String::new(),
    };
    if norad.is_empty() || !norad.chars().all(|c| c.is_ascii_digit()) {
        return Err(ApiError::validation("请输入有效的 NORAD 目录号（纯数字）"));
    }
    let nid: i64 = norad.parse().unwrap_or(0);
    let sats = store::load_satellites();
    if sats
        .iter()
        .any(|s| s.get("norad_id").and_then(|v| v.as_i64()) == Some(nid))
    {
        return Err(ApiError::validation(format!(
            "卫星（NORAD {}）已在已加入列表中",
            nid
        )));
    }
    let entry = libfiles::find_entry_by_norad(nid).ok_or_else(|| {
        ApiError::not_found(format!(
            "卫星库中未找到 NORAD {}（请先下载对应的数据源组）",
            nid
        ))
    })?;
    let name = entry
        .get("name")
        .and_then(|v| v.as_str())
        .filter(|s| !s.is_empty())
        .map(|s| s.to_string())
        .unwrap_or_else(|| norad.clone());
    let sat = json!({ "id": norad, "name": name, "norad_id": nid, "builtin": false });
    let mut next = sats.clone();
    next.push(sat.clone());
    let saved = store::save_satellites(&Value::Array(next));
    Ok(Json(json!({ "satellites": saved, "satellite": sat })))
}

/// POST /api/library/deactivate：把卫星从"已加入"列表移除（内置不可删除）。
pub async fn deactivate(Json(payload): Json<Value>) -> ApiResult<Json<Value>> {
    let sid = payload
        .get("id")
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .trim()
        .to_string();
    if sid.is_empty() {
        return Err(ApiError::validation("缺少卫星 id"));
    }
    let sats: Vec<Value> = store::load_satellites()
        .into_iter()
        .filter(|s| {
            let is_target = s.get("id").and_then(|v| v.as_str()) == Some(sid.as_str());
            let builtin = s.get("builtin").and_then(|v| v.as_bool()).unwrap_or(false);
            !(is_target && !builtin)
        })
        .collect();
    let saved = store::save_satellites(&Value::Array(sats));
    Ok(Json(json!({ "satellites": saved })))
}
